src/workloads/execution/ShardedLookup.yml

```diff
@@ -0,0 +1,118 @@
+    3. Fsync.
+    4. Running untargeted $lookup from a sharded collection.
+
+Actors:
+- Name: CreateShardedCollections
+  Type: AdminCommand
+  Threads: 1
+  Phases:
+  - Repeat: 1
+    Database: admin
+    Operations:
+    # Shard Collection0 and Collection1 using hashed sharding to ensure that the chunks are evenly
+    # distributed across the shards.
+    - OperationMetricsName: EnableSharding
+      OperationName: AdminCommand
+      OperationCommand:
+        enableSharding: &Database test
+    - OperationMetricsName: ShardLocalCollection
+      OperationName: AdminCommand
+      OperationCommand:
+        shardCollection: test.Collection0
+        key: {shardKey: hashed}
+        numInitialChunks: &NumChunks 6
+    - OperationMetricsName: ShardForeignCollection
+      OperationName: AdminCommand
+      OperationCommand:
+        shardCollection: test.Collection1
+        key: {shardKey: hashed}
+        numInitialChunks: *NumChunks
+    # Disable the balancer so that it can't skew results while the $lookups are running.
+    - OperationMetricsName: DisableBalancer
+      OperationName: AdminCommand
+      OperationCommand:
+        balancerStop: 1
+  - &Nop {Nop: true}
+  - *Nop
+  - *Nop
+
+- Name: LoadInitialData
+  Type: Loader
+  Threads: 1
+  Phases:
+  - *Nop
+  - Repeat: 1
+    BatchSize: 1000
+    Threads: 1
+    DocumentCount: &NumDocs 3000
+    Database: *Database
+    CollectionCount: 2    # Loader will populate 'Collection0' then 'Collection1'.
+    Document:
+      shardKey: {^RandomInt: {min: 1, max: 100}}
+      date: &Date {^RandomDate: {min: "2020-01-01", max: "2021-01-01"}}
+      int: {^RandomInt: {min: 1, max: 100}}
+  - *Nop
+  - *Nop
+
+- Name: Quiesce
+  Type: RunCommand
+  Threads: 1
+  Phases:
+  - *Nop
+  - *Nop
+  - Repeat: 1
+    Database: admin
+    Operations:
+    - OperationName: RunCommand
+      OperationCommand:
+        fsync: 1
+  - *Nop
+
+- Name: RunLookups
+  Type: RunCommand
+  Database: *Database
+  Phases:
+  - *Nop
+  - *Nop
+  - *Nop
+  - Repeat: 10    # Untargeted $lookup from sharded collection to sharded collection
+    Database: *Database
+    Operations:
+    - OperationMetricsName: UntargetedLookupShardedToSharded
+      OperationName: RunCommand
+      OperationCommand:
+        aggregate: Collection0
+        pipeline:
+          [{
+            $lookup: {
+              from: Collection1,
+              let: {localInt: "$int"},
+              pipeline: [{
+                $match: {
+                  $expr: {
+                    $and: [
+                      {$eq: ["$int", "$$localInt"]},
+                      {$lte: ["$date", *Date]}
+                    ]
+                  }
+                }
+              }],
+              as: matches
+            }
+          }]
+        # To get meaningful results, the entire result set should fit in a single batch. This should
+        # be possible since both collections are small.
+        cursor: {batchSize: *NumDocs}
```

Comment:
We need to make sure that the entire result set is calculated and returned in a single batch to get meaningful timing measurements. Whereas before we were seeing a linear increase in query time when increasing both collection sizes, now we see something that looks more quadratic (as expected for $lookup).

As a result, the queries are slower. Accordingly, I reduced the collection size and now each query takes about 11 seconds to run.

Reply:
Great, glad we were able to get to the bottom of those confounding scaling results!

As a back-of-the-envelope calculation, if each doc is about 0.7KB and each doc on the left side of the $lookup will match around 30 results (probably less) on the right, the total result set for each $lookup should never be more than 10MB, which is well below the 16MB limit. One thing to keep in mind when adding similar $graphLookup workloads is that the result set for a similar setup might be much larger than that.

Reply:
That makes sense, so this is lgtm.